### DISCUSSION TITLE
How to customize the Response?

### DISCUSSION THREAD

**Discussion by dlpuzcl** [vitalik/django-ninja#445] [Q&A] [answered]
n addition to the value of the schema in the JSON of the response, other fields need to be added.
Such as
```
    {  
          "code": 2000,
          "data": {
              "page": page,
              "limit": limit,
              "total": total,
              "data": data
          },
          "msg": msg
      }
```

**Comment by vitalik** [comment]
You might need to bring your full function code and goal details

**Reply by dlpuzcl** [reply]
Hi 
I want to extend HttpResponse and let it return more data，like this
```
from django.http import HttpResponse

class SuccessResponse(Response):

    def __init__(self, data=None, msg='success', status=None, template_name=None, headers=None, exception=False,
                 content_type=None,page=1,limit=1,total=1):
        std_data = {
            "code": 2000,
            "data": data
            "msg": msg
        }
        super().__init__(std_data, status, template_name, headers, exception, content_type)
```
If I use DRF, it can return `SuccessResponse(data=serializer.data, msg="Get data success")` directly.
I don't know if I've described it clearly, thank you!

**Comment by SmileyChris** [comment]
It looks like you're wanting some fancier response for paginated data. In MDN, I did this here: https://github.com/mdn/kuma/blob/main/kuma/api/v1/pagination.py

It's a bit to digest, but the key parts are to look at an implementation like here:
https://github.com/mdn/kuma/blob/main/kuma/api/v1/plus/notifications.py#L68

The key points are:
* the response is set to something like `LimitOffsetPaginatedResponse[NotificationSchema]` which is the layout of the reposnse
* the schema is decorated with `paginate(LimitOffsetPaginationWithMeta)` which handles processing the queryset data then returns `LimitOffsetPaginatedData` which formats everything into an object that the response expects.

The paginated response uses a generic so you can have the same base response class for different schema types, since the schema is the unique bit for different endpoints.

**Comment by dlpuzcl** [comment, marked as answer]
I implemented this function by override create_response(), thanks for SmileyChris, and vitalik  help.
```
class FuNinjaAPI(NinjaAPI):
	def create_response(
			self, request: HttpRequest, data: Any, *, status: int = 200, msg: str = "success"
	) -> HttpResponse:
		std_data = {
			"code": status,
			"data": data,
			"msg": msg
		}
		content = self.renderer.render(request, std_data, response_status=status)
		content_type = "{}; charset={}".format(
			self.renderer.media_type, self.renderer.charset
		)

		return HttpResponse(content, status=status, content_type=content_type)
```

**Reply by johnbridstrup** [reply]
This does mean the api schema that is hosted does not match the final response structure, yes?

**Reply by sebastian-philipp** [reply]
Agree. Some people rely on the openapi.json to match the actual output. Especially those, who use code generators to generate client code.
